compiler/rustc_infer/src/infer/relate/equate.rs

```diff
@@ -159,8 +160,22 @@ impl<'tcx> TypeRelation<'tcx> for Equate<'_, '_, 'tcx> {
         }
 
         if a.skip_binder().has_escaping_bound_vars() || b.skip_binder().has_escaping_bound_vars() {
-            self.fields.higher_ranked_sub(a, b, self.a_is_expected)?;
-            self.fields.higher_ranked_sub(b, a, self.a_is_expected)?;
+            // Check whether `for<..> A == for<..> B` holds.
+            // For this to hold, we need to check `for<..> A <: for<..> B` and
+            // `for<..> B <: for<..> A` in an invariant context.
+
+            let span = self.fields.trace.cause.span;
+            let infcx = self.fields.infcx;
+
+            // Check if `for<..> A <: for<..> B` holds.
+            let b_prime = infcx.instantiate_binder_with_placeholders(b);
+            let a_prime = infcx.instantiate_binder_with_fresh_vars(span, HigherRankedType, a);
+            self.relate(a_prime, b_prime)?;
+
+            // Check if `for<..> B <: for<..> A` holds.
+            let a_prime = infcx.instantiate_binder_with_placeholders(a);
+            let b_prime = infcx.instantiate_binder_with_fresh_vars(span, HigherRankedType, b);
+            self.fields.equate(!self.a_is_expected).relate(b_prime, a_prime)?;
```

Comment:
Maybe introduce `with_expected_switched` here, like `Sub::with_expected_switched` does.

Reply:
I'd leave @lcnr's change for a follow up to avoid mixing things with this fix.